### PR TITLE
single api framework

### PR DESCRIPTION
**Pull request by TeganAli**
# About

This is a post about...

# Pre-requisites

- [ ] The draft is complete and this post is ready to be reviewed.
- [ ] I'm familiar with the [writing for Octopus TL;DR](https://style.octopus.com/writing-at-octopus-tldr).
- [ ] The PR build passes validation, and if it doesn't, I've checked the [common validations errors](https://style.octopus.com/writing-at-octopus-tldr#common-validation-errors) and none of those apply.
- [ ] The images I've included follow the [rules for working with images](https://style.octopus.com/images#rules-for-screenshots).

### CODE REVIEW

File: blog/2021-08/single-api-framework/index.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+We used [Assent](https://github.com/droyad/Assent) tests to snapshot the entire HTTP response from an endpoint, for both the happy path and any edge cases. These tests also measured the response time for an endpoint so we could check for any performance regressions.
+
+### Migrate
+
+For a simple GET endpoint, the migration could be trivial. For endpoints which mutated state, things could be more time consuming. Rules such as validation were considered, understood, and ported over to the ASP.NET controller pattern.
+
+### Verify
+
+The previously-added tests were re-run in CI, alerting us to any performance regressions, functional changes or alterations of the response content. Objective decisions were then made about changes and whether they were acceptable or not.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:49]:
```suggestion
The previously-added tests were re-run in CI, alerting us to any performance regressions, functional changes, or alterations to the response content. Objective decisions were then made about changes and whether they were acceptable or not.
```

File: blog/2021-08/single-api-framework/index.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+To refine migrations, we targeted endpoints which were similar and less-complex. To find out which endpoints met these criteria we wrote a C# script, executed in LINQPad, which reflected over the Octopus Deploy codebase and grouped endpoints by common code and base classes. 
+
+To get a measure of relative complexity, we sorted by a metric based on lines of code and cyclomatic complexity. This grouping and sorting allowed us to focus on the mechanism of migration while avoiding too much domain complexity.
+
+We learned a few things:
+
+- Due to the performance improvements of ASP.NET, it wasn't necessary to continue testing for performance regressions.
+- Snapshot-based tests can be challenging to write because you have to account for and “scrub” data that changes between invocations of an endpoint, such as timestamps or IDs.
+- For tests that are checking edge cases, it can be simpler to just assert on the parts of the response that you really care about, such as response code and error text.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:70]:
```suggestion
- For tests that are checking edge cases, it can be simpler to assert on the parts of the response that you really care about, such as response code and error text.
```

File: blog/2021-08/single-api-framework/index.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+The previously-added tests were re-run in CI, alerting us to any performance regressions, functional changes, or alterations to the response content. Objective decisions were then made about changes and whether they were acceptable or not.
+
+## Enable
+
+This approach was successful and proved that migrating the entire Octopus Deploy API was possible. 
+
+It also showed us:
+
+- Nancy endpoints typically took around 100ms to return a response.
+- The first call to an ASP.NET endpoint would return in around 200ms but subsequent calls took only 30 – 40ms.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:58]:
```suggestion
- The first call to an ASP.NET endpoint would return in around 200ms but subsequent calls took only 30–40ms.
```

File: blog/2021-08/single-api-framework/index.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+2. Migrate
+3. Verify
+
+### Analyse & Test
+
+We used [Assent](https://github.com/droyad/Assent) tests to snapshot the entire HTTP response from an endpoint, for both the happy path and any edge cases. These tests also measured the response time for an endpoint so we could check for any performance regressions.
+
+### Migrate
+
+For a simple GET endpoint, the migration could be trivial. For endpoints which mutated state, things could be more time consuming. Rules such as validation were considered, understood, and ported over to the ASP.NET controller pattern.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:45]:
```suggestion
For a simple GET endpoint, the migration could be trivial. For endpoints that mutated state, things could be more time consuming. Rules such as validation were considered, understood, and ported over to the ASP.NET controller pattern.
```

File: blog/2021-08/single-api-framework/index.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+To get a measure of relative complexity, we sorted by a metric based on lines of code and cyclomatic complexity. This grouping and sorting allowed us to focus on the mechanism of migration while avoiding too much domain complexity.
+
+We learned a few things:
+
+- Due to the performance improvements of ASP.NET, it wasn't necessary to continue testing for performance regressions.
+- Snapshot-based tests can be challenging to write because you have to account for and “scrub” data that changes between invocations of an endpoint, such as timestamps or IDs.
+- For tests checking edge cases, it's simpler to assert on the parts of the response that you really care about, such as response code and error text.
+- Tests need data, which can be nontrivial to create, so it’s important to separate this concern from the specifics of your tests.
+- It's vital to establish (and document) repeatable patterns for how you migrate, so others in the engineering team can do migrations as well.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:72]:
```suggestion
- It's vital to establish and document repeatable patterns for how you migrate, so others in the engineering team can do migrations as well.
```

File: blog/2021-08/single-api-framework/index.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+This approach leverages developer domain knowledge and spreads the migration work across the engineering team.
+
+
+## Conclusion
+
+- A migration like this can be done, but it's a significant investment of time.
+- It’s important to quantify the work, and find the most effective way to organize efforts.
+- This is very parallelizable work.
+- Establishing (and documenting) patterns allows any developer in your organization to help.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:96]:
```suggestion
- Establishing and documenting patterns allows any developer in your organization to help.
```

File: blog/2021-08/single-api-framework/index.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+2. Migrate
+3. Verify
+
+### Analyse & Test
+
+We used [Assent](https://github.com/droyad/Assent) tests to snapshot the entire HTTP response from an endpoint, for both the happy path and any edge cases. These tests also measured the response time for an endpoint so we could check for any performance regressions.
+
+### Migrate
+
+For a simple GET endpoint, the migration could be trivial. For endpoints that mutated state, things could be more time consuming. Rules such as validation were considered, understood, and ported over to the ASP.NET controller pattern.
```
Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:45]: @leecarlon is the word "the" missing here, should it say: “For endpoints that mutated **the** state..." ?

Comment (TeganAli) [inline blog/2021-08/single-api-framework/index.md:45]:
```suggestion
For a simple GET endpoint, the migration could be trivial. For endpoints that mutated the state, things could be more time consuming. Rules such as validation were considered, understood, and ported over to the ASP.NET controller pattern.
```

File: blog/2021-08/single-api-framework/index.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
+---
+title: A single API framework
+description: We migrated Octopus Deploy to a more standardized way of writing an API, using ASP.NET controllers. Learn about our implementation strategy: Derisk, Enable, Finish.
```
Comment (wordlee) [inline blog/2021-08/single-api-framework/index.md:3]:
```suggestion
description: "We migrated Octopus Deploy to a more standardized way of writing an API, using ASP.NET controllers. Learn about our implementation strategy: Derisk, Enable, Finish."
```
